### PR TITLE
[BEAM-2193] implement join

### PR DESCRIPTION
**Pull request by xumingming**
implement join

### CODE REVIEW

**Comment by aaltay** [comment]
R: @XuMingmin @takidau

File: dsls/pom.xml [hunk 1]
```diff
@@ -22,7 +22,7 @@
   <parent>
     <groupId>org.apache.beam</groupId>
     <artifactId>beam-parent</artifactId>
-    <version>0.7.0-SNAPSHOT</version>
+    <version>2.1.0-SNAPSHOT</version>
```
Comment (takidau) [inline dsls/pom.xml]: I would rather not conflate upgrading the underlying version w/ adding join support. Can that be done in a separate PR, please?

File: dsls/sql/src/main/java/org/apache/beam/dsls/sql/rel/BeamJoinRel.java [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+import org.apache.calcite.rel.core.JoinRelType;
+import org.apache.calcite.rex.RexCall;
+import org.apache.calcite.rex.RexInputRef;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.util.Pair;
+
+/**
+ * {@code BeamRelNode} to replace a {@code Join} node.
+ *
+ * <p>Support for join can be categorized into 3 cases:
```
Comment (takidau) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/rel/BeamJoinRel.java]: We will need to be careful here. This is a perfectly reasonable initial set of supported use cases, but I think there will be more variations we want to support in the future. As long as SQL remains in a branch, we can probably forge ahead with this change, but I will want to make sure we have a comprehensive plan defined for JOIN support before merging to master. Let me try to get my SQL spec doc finished soon, that will hopefully give us a good place to discuss what we want to pursue.

File: dsls/sql/src/main/java/org/apache/beam/dsls/sql/rel/BeamJoinRel.java [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+    final int rightIndex = rightIndex1 - separator;
+
+    return new Pair<>(leftIndex, rightIndex);
+  }
+
+  // NOTE: the main piece of this method is copied from the joinlibrary
+  //       currently joinlibrary does not support FULL OUTER JOIN
+  //       once FULL OUTER JOIN is supported in joinlibrary we will use
+  //       joinlibrary instead.
+  //       Pull Request: https://github.com/apache/beam/pull/3267
```
Comment (takidau) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/rel/BeamJoinRel.java]: It looks like this PR is merged, correct? Please resolve this if so.

File: dsls/pom.xml (hunk 1 above)
Comment (xumingming) [inline dsls/pom.xml]: I will remove this change.

File: dsls/sql/src/main/java/org/apache/beam/dsls/sql/rel/BeamJoinRel.java (hunk 2 above)
Comment (xumingming) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/rel/BeamJoinRel.java]: Yeah, looking forward to your spec :)

File: dsls/sql/src/main/java/org/apache/beam/dsls/sql/rel/BeamJoinRel.java (hunk 3 above)
Comment (xumingming) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/rel/BeamJoinRel.java]: Will do this after rebase from DSL_SQL.

**Comment by xumingming** [comment]
Rebase done.

**Comment by xumingming** [comment]
Rebase done.

File: dsls/sql/src/test/java/org/apache/beam/dsls/sql/planner/MockedBeamSqlTable.java [hunk 4]
```diff
@@ -49,6 +48,7 @@
   public static final ConcurrentLinkedQueue<BeamSqlRow> CONTENT = new ConcurrentLinkedQueue<>();
 
   private List<BeamSqlRow> inputRecords;
+  private PCollection.IsBounded isBounded = PCollection.IsBounded.BOUNDED;
```
Comment (XuMingmin) [inline dsls/sql/src/test/java/org/apache/beam/dsls/sql/planner/MockedBeamSqlTable.java]: not sure if this can test both `BOUNDED` and `UNBOUNDED` case, as `Create.of()` returns a BOUNDED PCollection.  Could use `TestStream` IMO.

**Comment by XuMingmin** [comment]
@takidau I'm thinking to finish this PR first as an initial implementation of `JOIN`, to be included in the merge_to_master thread.  We could keep enhance to align with your proposal specs, I'm sure there's much space to improve.

**Comment by takidau** [comment]
I'm going to dive into this tomorrow (no more training :-). I basically want to make sure we don't introduce anything we might want to roll back in the future semantics-wise.

File: dsls/sql/pom.xml [hunk 5]
```diff
@@ -149,6 +149,11 @@
       <scope>provided</scope>
     </dependency>
     <dependency>
+      <groupId>org.apache.kafka</groupId>
```
Comment (xumingming) [inline dsls/sql/pom.xml]: Should be removed.

File: dsls/sql/src/test/java/org/apache/beam/dsls/sql/planner/MockedBeamSqlTable.java (hunk 4 above)
Comment (xumingming) [inline dsls/sql/src/test/java/org/apache/beam/dsls/sql/planner/MockedBeamSqlTable.java]: The core feature of `TestStream` is enable tests to advance the watermark and processing time as elements are emitted. Which seems not needed for our test? We just need `IsBounded` to tell whether specified tables can join?

Comment (XuMingmin) [inline dsls/sql/src/test/java/org/apache/beam/dsls/sql/planner/MockedBeamSqlTable.java]: I don't know if you can change the BOUNDED/UNBOUNDED behavior of PCollection by set it here. The test for UNBOUNDED is more about triggers/watermark to me.

Comment (xumingming) [inline dsls/sql/src/test/java/org/apache/beam/dsls/sql/planner/MockedBeamSqlTable.java]: @XuMingmin make sense. Let me add a `MockedUnboundedTable` and add real test for unbounded table.

**Comment by xumingming** [comment]
Update: 

1. Added a `MockedUnboundedTable` which use `TestStream` to mock a _real_ unbounded table.
1. Updated the JOIN test to use `MockedUnboundedTable`.

**Comment by xumingming** [comment]
Since now we have `MockedBeamSqlTable`(might better be renamed to MockedBoundedTable) and `MockedUnboundedTable`, some refactoring is needed to cleanup the mock logic and related tests, created https://issues.apache.org/jira/browse/BEAM-2515 to track it.

**Comment by xumingming** [comment]
you mean `kafka-clients`, right? Seems it is actually needed. Our pom has configured the maven dependency plugin to check whether there are unused dependency, and whether there are dependency which are used but not declared, so the code can build, the dependencies should be correct, we seems not need to review the pom manually?

**Comment by XuMingmin** [comment]
Make sense, let's keep it.

File: dsls/sql/src/main/java/org/apache/beam/dsls/sql/rel/BeamJoinRel.java [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+  private BeamSqlRow buildNullRow(BeamRelNode relNode) {
+    BeamSqlRecordType leftType = CalciteUtils.toBeamRecordType(relNode.getRowType());
+    BeamSqlRow nullRow = new BeamSqlRow(leftType);
+    for (int i = 0; i < leftType.size(); i++) {
+      nullRow.addField(i, null);
+    }
+    return nullRow;
+  }
+
+  private List<Pair<Integer, Integer>> extractJoinColumns(int separator) {
```
Comment (takidau) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/rel/BeamJoinRel.java]: What is separator here? Is it just an offset used to effectively remove any columns on the left side? If so, it could maybe use a better name.

File: dsls/sql/src/main/java/org/apache/beam/dsls/sql/transform/BeamJoinTransforms.java [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+
+
+  /**
+   * A {@code DoFn} which implement the sideInput-JOIN.
+   */
+  public static class SideInputJoinDoFn extends DoFn<KV<BeamSqlRow, BeamSqlRow>, BeamSqlRow> {
+    private PCollectionView<Map<BeamSqlRow, Iterable<BeamSqlRow>>> sideInputView;
+    private JoinRelType joinType;
+    private BeamSqlRow rightNullRow;
+    private boolean swap;
```
Comment (takidau) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/transform/BeamJoinTransforms.java]: Final for all?

File: dsls/sql/src/test/java/org/apache/beam/dsls/sql/TestUtils.java [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+
+    return strs;
+  }
+
+  /**
+   * Convenient way to build a list of {@code BeamSqlRow}s.
+   */
+  public static class RowsBuilder {
+    private BeamSqlRecordType type;
+    private List<BeamSqlRow> rows = new ArrayList<>();
```
Comment (takidau) [inline dsls/sql/src/test/java/org/apache/beam/dsls/sql/TestUtils.java]: Both final? I think you'll have to add a constructor, but it's the proper way to do it unless type needs to be mutable.

File: dsls/sql/src/test/java/org/apache/beam/dsls/sql/planner/MockedUnboundedTable.java [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+import org.apache.beam.sdk.values.TimestampedValue;
+import org.apache.calcite.util.Pair;
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+
+/**
+ * A mocked unbounded table.
+ */
+public class MockedUnboundedTable extends MockedTable {
+  private List<Pair<Duration, List<BeamSqlRow>>> timestampedRows = new ArrayList<>();
```
Comment (takidau) [inline dsls/sql/src/test/java/org/apache/beam/dsls/sql/planner/MockedUnboundedTable.java]: final?

File: dsls/sql/src/main/java/org/apache/beam/dsls/sql/rel/BeamJoinRel.java [hunk 10]
```diff
@@ -0,0 +1,275 @@
...
+        && extractedLeftRows.isBounded() == PCollection.IsBounded.BOUNDED;
+
+    PCollection<KV<BeamSqlRow, BeamSqlRow>> realLeftRows =
+        swapped ? extractedRightRows : extractedLeftRows;
+    PCollection<KV<BeamSqlRow, BeamSqlRow>> realRightRows =
+        swapped ? extractedLeftRows : extractedRightRows;
+    BeamSqlRow realRightNullRow = swapped ? leftNullRow : rightNullRow;
+    JoinRelType realJoinType = swapped ? JoinRelType.LEFT : joinType;
+
+    final PCollectionView<Map<BeamSqlRow, Iterable<BeamSqlRow>>> rowsView = realRightRows
```
Comment (takidau) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/rel/BeamJoinRel.java]: Factor everything below here into a helper function that's called with the rows already swapped if necessary? Will make reading the code easier and less error prone.

File: dsls/sql/src/main/java/org/apache/beam/dsls/sql/rel/BeamJoinRel.java (hunk 2 above)
Comment (takidau) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/rel/BeamJoinRel.java]: Thanks for your patience. The spec proposal is not quite ready yet, but close. At this point though, I think what you have should be compatible with what I'm proposing. There will be a separate, but related, question of default triggering, but we can tackle that as part of the merge burndown list (added there).

File: dsls/sql/src/main/java/org/apache/beam/dsls/sql/rel/BeamJoinRel.java [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+ * <ul>
+ *   <li>BoundedTable JOIN BoundedTable</li>
+ *   <li>UnboundedTable JOIN UnboundedTable</li>
+ *   <li>BoundedTable JOIN UnboundedTable</li>
+ * </ul>
+ *
+ * <p>For the first two cases, a standard join can be utilized to implement them as long as the
+ * windowFn of the both sides match. For the third case, {@code sideInput} is utilized to implement
+ * the join, hence there are some constrains for the third case: 1) FULL JOIN is not supported
+ * 2) The unbounded table must be at the left side of the OUTER JOIN.
```
Comment (takidau) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/rel/BeamJoinRel.java]: The left side constraint is an implementation detail, isn't it? It appears you automatically swap left and right sides internally for the sake of doing the join then swap the results back if the unbounded table is on the right, don't you?

File: dsls/sql/src/main/java/org/apache/beam/dsls/sql/rel/BeamJoinRel.java [hunk 12]
```diff
@@ -0,0 +1,275 @@
...
+ *
+ * <p>There is also some overall constrains:
```
Comment (takidau) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/rel/BeamJoinRel.java]: constraints

File: dsls/sql/src/main/java/org/apache/beam/dsls/sql/rel/BeamJoinRel.java [hunk 13]
```diff
@@ -0,0 +1,275 @@
...
+ *
+ * <ul>
+ *  <li>Only equi-join is supported</li>
+ *  <li>CROSS JOIN is not supported</li>
```
Comment (takidau) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/rel/BeamJoinRel.java]: Should there be tests verifying attempts to violate these constraints fail gracefully?

File: dsls/sql/src/main/java/org/apache/beam/dsls/sql/transform/BeamJoinTransforms.java [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+  }
+
+  private static BeamSqlRow combineTwoRowsIntoOne(BeamSqlRow leftRow, BeamSqlRow rightRow) {
+    return combineTwoRowsIntoOne(leftRow, rightRow, false);
+  }
+
+  /**
+   * As the method name suggests: combine two rows into one wide row.
+   */
+  private static BeamSqlRow combineTwoRowsIntoOne(BeamSqlRow leftRow,
```
Comment (xumingming) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/transform/BeamJoinTransforms.java]: Good point!

File: dsls/sql/src/test/java/org/apache/beam/dsls/sql/rel/BeamJoinRelUnboundedVsBoundedTest.java [hunk 15]
```diff
@@ -0,0 +1,216 @@
...
+        .of(SqlTypeName.INTEGER, "order_id",
+            SqlTypeName.VARCHAR, "buyer",
+
+            1, "james",
+            2, "bond"
+        ));
+  }
+
+  @Test
+  public void testInnerJoin() throws Exception {
```
Comment (xumingming) [inline dsls/sql/src/test/java/org/apache/beam/dsls/sql/rel/BeamJoinRelUnboundedVsBoundedTest.java]: Good point, find a related bug.

File: dsls/sql/src/test/java/org/apache/beam/dsls/sql/rel/BeamJoinRelUnboundedVsUnboundedTest.java [hunk 16]
```diff
@@ -0,0 +1,222 @@
...
+        + "          GROUP BY order_id, TUMBLE(order_time, INTERVAL '1' HOUR)) o2 "
+        + " on "
+        + " o1.order_id=o2.order_id"
+        ;
+
+    // 1, 1 | 1, 3
+    // 2, 2 | NULL, NULL
+    // ---- | -----
+    // 2, 2 | 2, 5
+    // 3, 3 | NULL, NULL
```
Comment (xumingming) [inline dsls/sql/src/test/java/org/apache/beam/dsls/sql/rel/BeamJoinRelUnboundedVsUnboundedTest.java]: Yeah, I missed this one, will fix.

File: dsls/sql/src/test/java/org/apache/beam/dsls/sql/planner/MockedBeamSqlTable.java [hunk 17]
```diff
@@ -119,10 +117,11 @@ public RelDataType apply(RelDataTypeFactory a0) {
 
   @Override
   public BeamIOType getSourceType() {
-    return BeamIOType.UNBOUNDED;
+    return BeamIOType.BOUNDED;
```
Comment (xumingming) [inline dsls/sql/src/test/java/org/apache/beam/dsls/sql/planner/MockedBeamSqlTable.java]: I intended to make this change in another PR(to avoid unrelated changes). Created a separate issue to track this: https://issues.apache.org/jira/browse/BEAM-2515

File: dsls/sql/src/test/java/org/apache/beam/dsls/sql/planner/MockedTable.java [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+package org.apache.beam.dsls.sql.planner;
+
+import java.util.concurrent.atomic.AtomicInteger;
+import org.apache.beam.dsls.sql.schema.BaseBeamTable;
+import org.apache.beam.dsls.sql.schema.BeamSqlRecordType;
+
+/**
+ * Base class for mocked table.
+ */
+public abstract class MockedTable extends BaseBeamTable {
```
Comment (xumingming) [inline dsls/sql/src/test/java/org/apache/beam/dsls/sql/planner/MockedTable.java]: ditto, will do the change in a dedicated PR.

File: dsls/sql/src/test/java/org/apache/beam/dsls/sql/planner/MockedUnboundedTable.java [hunk 19]
```diff
@@ -0,0 +1,120 @@
...
+import org.apache.beam.sdk.values.PDone;
+import org.apache.beam.sdk.values.TimestampedValue;
+import org.apache.calcite.util.Pair;
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+
+/**
+ * A mocked unbounded table.
+ */
+public class MockedUnboundedTable extends MockedTable {
```
Comment (xumingming) [inline dsls/sql/src/test/java/org/apache/beam/dsls/sql/planner/MockedUnboundedTable.java]: ditto.

File: dsls/sql/src/test/java/org/apache/beam/dsls/sql/planner/MockedUnboundedTable.java (hunk 9 above)
Comment (xumingming) [inline dsls/sql/src/test/java/org/apache/beam/dsls/sql/planner/MockedUnboundedTable.java]: The `timestampRows` are configured step by step, so can't be set to `final`.

File: dsls/sql/src/main/java/org/apache/beam/dsls/sql/rel/BeamJoinRel.java (hunk 11 above)
Comment (xumingming) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/rel/BeamJoinRel.java]: Yes, I have done the auto-swap. The javadoc should be updated.

Comment (xumingming) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/rel/BeamJoinRel.java]: Actually there is still a constraint, unbounded table should on the left side of the join after auto-swap, e.g. we don't support the follow join:

```sql
select * from bounded_table LEFT OUTER JOIN unbounded_table.
```

File: dsls/sql/src/main/java/org/apache/beam/dsls/sql/rel/BeamJoinRel.java (hunk 13 above)
Comment (xumingming) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/rel/BeamJoinRel.java]: Will do.

File: dsls/sql/src/main/java/org/apache/beam/dsls/sql/rel/BeamJoinRel.java (hunk 6 above)
Comment (xumingming) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/rel/BeamJoinRel.java]: rename to `leftRowColumnCount`.

File: dsls/sql/src/main/java/org/apache/beam/dsls/sql/rule/BeamJoinRule.java [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+import org.apache.calcite.rel.core.Join;
+import org.apache.calcite.rel.logical.LogicalJoin;
+
+/**
+ * {@code ConverterRule} to replace {@code Join} with {@code BeamJoinRel}.
+ */
+public class BeamJoinRule extends ConverterRule {
+  public static final BeamJoinRule INSTANCE = new BeamJoinRule();
+  private BeamJoinRule() {
+    super(LogicalJoin.class, Convention.NONE,
```
Comment (xumingming) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/rule/BeamJoinRule.java]: because the first param of `ConverterRule`(`BeamJoinRule`'s parent) is(I quote):

```
Type of relational expression to consider converting
```

For a join, before convert, it's relation algebra tree is something like:

```bash
LogicalProject(order_id=[$0], sum_site_id=[$1], order_id0=[$2], sum_site_id0=[$3])
  LogicalJoin(condition=[=($0, $2)], joinType=[inner])
    LogicalProject(order_id=[$0], sum_site_id=[$2])
      LogicalAggregate(group=[{0, 1}], sum_site_id=[SUM($2)])
        LogicalProject(order_id=[$0], $f1=[TUMBLE($3, 3600000)], site_id=[$1])
          LogicalTableScan(table=[[ORDER_DETAILS]])
    LogicalProject(order_id=[$0], sum_site_id=[$2])
      LogicalAggregate(group=[{0, 1}], sum_site_id=[SUM($2)])
        LogicalProject(order_id=[$0], $f1=[TUMBLE($3, 3600000)], site_id=[$1])
          LogicalTableScan(table=[[ORDER_DETAILS]])
```

Note the `LogicalJoin` in the tree, it is the class we are trying to convert to `BeamJoinRel`.

**Comment by xumingming** [comment]
Updated the code according to review comments @takidau

File: dsls/sql/src/main/java/org/apache/beam/dsls/sql/rule/BeamJoinRule.java (hunk 20 above)
Comment (takidau) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/rule/BeamJoinRule.java]: Makes sense, thx for the detailed explanation. 👍

File: dsls/sql/src/test/java/org/apache/beam/dsls/sql/planner/MockedBeamSqlTable.java (hunk 17 above)
Comment (takidau) [inline dsls/sql/src/test/java/org/apache/beam/dsls/sql/planner/MockedBeamSqlTable.java]: sgtm, thx

File: dsls/sql/src/main/java/org/apache/beam/dsls/sql/rel/BeamJoinRel.java [hunk 21]
```diff
@@ -242,17 +266,22 @@ private BeamSqlRow buildNullRow(BeamRelNode relNode) {
     return nullRow;
   }
 
-  private List<Pair<Integer, Integer>> extractJoinColumns(int separator) {
+  private List<Pair<Integer, Integer>> extractJoinColumns(int leftRowColumnCount) {
+    // it's a CROSS JOIN because: condition == true
+    if (condition instanceof RexLiteral && (Boolean) ((RexLiteral) condition).getValue()) {
```
Comment (takidau) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/rel/BeamJoinRel.java]: For my own edification, can you explain this a little further? What is the condition here, and why does it being set to true imply CROSS JOIN?

**Comment by takidau** [comment]
Merged.

File: dsls/sql/src/main/java/org/apache/beam/dsls/sql/rel/BeamJoinRel.java (hunk 21 above)
Comment (xumingming) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/rel/BeamJoinRel.java]: For the following cross join:

```sql
select * from person p1, person p2
```

its relation algebra tree is:

```
LogicalProject(id=[$0], name=[$1], age=[$2], ts=[$3], boolean_c=[$4], id0=[$5], name0=[$6], age0=[$7], ts0=[$8], boolean_c0=[$9])
  LogicalJoin(condition=[true], joinType=[inner])
    EnumerableTableScan(table=[[person]])
    EnumerableTableScan(table=[[person]])
```

so in Calcite, a `CROSS JOIN` is converted to a `INNER JOIN` with the `condition` set to `true`.

**Comment by xumingming** [comment]
Thanks @takidau @XuMingmin
